chore: wrap relevant surfaces with the contexts needed to dispatch alert modal

dblandin opened the pull request:
The type of this PR is: **Chore**

<!-- Build / Chore / CI / Docs / Feat / Fix / Perf / Refactor / Revert / Style / Test -->

<!-- If applicable, write the Jira ticket number in square brackets e.g. `[GRO-434]`
     The Jira integration will turn it into a clickable link for you. -->

### Description

Wrap relevant surfaces with the contexts needed to dispatch alert modal.


[GRO-434]: https://artsyproduct.atlassian.net/browse/GRO-434?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

damassi commented on src/Apps/Artist/Routes/WorksForSale/Components/ArtistArtworkFilter.tsx:
```diff
@@ -67,16 +68,22 @@ const ArtistArtworkFilter: React.FC<ArtistArtworkFilterProps> = props => {
-          }}
-        />
-      </ArtworkFilterSavedSearchAlertContextProvider>
+      <ArtworkFilterAlertContextProvider
```
Curious why everything needs to be wrapped all around the app, vs placing it once in BaseArtworkFilter, or wrapping a component above that (but yet internal to it)

dblandin commented on src/Apps/Artist/Routes/WorksForSale/Components/ArtistArtworkFilter.tsx:
```diff
@@ -67,16 +68,22 @@ const ArtistArtworkFilter: React.FC<ArtistArtworkFilterProps> = props => {
-      </ArtworkFilterSavedSearchAlertContextProvider>
+      <ArtworkFilterAlertContextProvider
+        initialCriteria={{ artistIDs: [artist.internalID] }}
+      >
```
We're wrapping here because there are quite a few components nested under this point in the component tree that render a Create Alert button and each needs to be able to dispatch the modal.

dblandin commented on src/Apps/Artist/Routes/WorksForSale/Components/ArtistWorksForSaleEmpty.tsx:
```diff
@@ -47,10 +41,11 @@ const ArtistWorksForSaleEmpty: FC<ArtistWorksForSaleEmptyProps> = ({
-              {alertComponent}
-            </>
+            <ArtworkFilterAlertContextProvider
+              initialCriteria={{ artistIDs: [artist.internalID] }}
```
Initial criteria including only `artistIDs` but this is refined via the artwork grid filter store further down the stack.

dblandin commented on src/Apps/Artwork/ArtworkApp.tsx:
```diff
@@ -292,6 +294,12 @@ const TrackingWrappedArtworkApp: React.FC<Props> = props => {
+    attributionClass: compact([attributionClass?.internalID]),
+    artistIDs: compact(artists).map(artist => artist.internalID),
+    additionalGeneIDs: [mediumType?.filterGene?.slug as string],
+  }
```
Initial criteria is a bit more involved here including Medium Type and Rarity values.

dblandin commented on src/Components/ArtworkFilter/ArtworkFilterCreateAlert.tsx:
```diff
@@ -23,55 +21,45 @@ export const ArtworkFilterCreateAlert: FC<ArtworkFilterCreateAlertProps> = ({
-
-  const { clickedCreateAlert } = useAlertTracking()
+  // If there is no entity then we don't want to create an alert
+  if (isEmpty(entity)) return null
```
Still relying on this entity concept to conditionally render on the right pages, but I'm hoping we can refactor this out soon.

olerichter00 commented on src/Apps/Artwork/ArtworkApp.tsx:
```diff
@@ -292,6 +294,12 @@ const TrackingWrappedArtworkApp: React.FC<Props> = props => {
+  const initialAlertCriteria = {
+    attributionClass: compact([attributionClass?.internalID]),
+    artistIDs: compact(artists).map(artist => artist.internalID),
+    additionalGeneIDs: [mediumType?.filterGene?.slug as string],
```
Could we use `compact` here as well to filter out `undefined` and `null` values instead of the type assertion `as string`?

dblandin replied:
Ahh, that's a good point. Will update 👍🏼

olerichter00 commented on src/Apps/Artwork/Components/ArtworkSidebar/ArtworkSidebarBiddingClosedMessage.tsx:
```diff
@@ -43,13 +33,8 @@ const BiddingClosedMessage: React.FC<BiddingClosedMessageProps> = ({
-                />
+                <CreateAlertButton width="100%" size="large" />
               </ProgressiveOnboardingAlertCreateSimple>
-              {alertComponent}
```
**question:** Are you removing `alertComponent` here because it's already present in the Sidebar?

dblandin replied:
Instead of explicitly rendering this modal/alert component next to each `Create Alert` button the idea is to place an AlertProvider component at a point in the component tree which 1) has access to the desired initialCriteria 2) renders one instance of the modal component alongside its children 3) will wrap any other branches of the component tree which might need to dispatch the modal via (Create Alert) buttons.

So this PR hoists the modal component up not only from the sidebar but from any deeper spot in the component tree.

olerichter00 replied:
Great, thanks for the explanation 🙏

olerichter00 commented on src/Components/Alert/Hooks/useAlertContext.tsx:
```diff
@@ -163,23 +177,37 @@ export const AlertProvider: FC<AlertProviderProps> = ({
-  const { createdAlert } = useAlertTracking()
+  useEffect(() => {
+    const criteria =
+      getAllowedSearchCriteria(initialCriteria as SearchCriteriaAttributes) ??
```
As this is equivalent to the now forbidden non-null assertion (https://github.com/artsy/force/pull/13083), we should probably remove the assertion. For example, like this:

```suggestion
      getAllowedSearchCriteria(initialCriteria || {}) ??
```

olerichter00 commented on src/Components/Alert/Hooks/useAlertContext.tsx:
```diff
@@ -163,23 +177,37 @@ export const AlertProvider: FC<AlertProviderProps> = ({
+      email: true,
+    },
+    criteria: getAllowedSearchCriteria(
+      initialCriteria as SearchCriteriaAttributes
```
Same here:

```suggestion
      initialCriteria || {}
```

olerichter00 reviewed (approved):
Looks good 👍

dblandin commented on src/Components/Alert/Hooks/__tests__/AlertProvider.jest.tsx:
```diff
@@ -0,0 +1,121 @@
+    relayEnv.mockClear()
+  })
+
+  it("opens the create alert modal with initial criteria", async () => {
```
@tam-kis This should give us a good base to continue iterating on. Will add some additional specs exercising the filter option in a follow-up PR.

dblandin commented on src/Components/Alert/AlertProvider.tsx:
```diff
@@ -0,0 +1,208 @@
+    })
+
+    return () => {
+      subscription.unsubscribe()
```
@olerichter00 Hopefully this reduces the chance of any runaway network requests causing trouble after the component dismounts.

damassi commented on src/Components/Alert/Hooks/__tests__/AlertProvider.jest.tsx:
```diff
@@ -0,0 +1,121 @@
+
+  beforeAll(() => {
+    trackEvent = jest.fn()
+    ;(useTracking as jest.Mock).mockImplementation(() => ({ trackEvent }))
```
`mockTrackEvent = trackEvent as jest.Mock`
`mockUseSystemContext = ...` 
`mockUseAuthDialog`

dblandin replied:
Thanks! Hopefully [this](https://github.com/artsy/force/pull/13153/commits/9cd7fe1b7b2d2e8c81297250ccd719b234e55fb4) is what you mean. Definitely looks a bit cleaner.

damassi commented on src/Components/Alert/AlertContext.ts:
```diff
@@ -0,0 +1,148 @@
+  goToFilters: () => {},
+  onComplete: () => {},
+  state: DEFAULT_STATE,
+})
```
If you don't feel like setting default functions and stick specifically to state -- that kind of thing -- can type as:
```
} as unknown as AlertContextProps)
```

damassi commented on src/Components/Alert/Components/CreateAlertButton.tsx:
```diff
@@ -3,22 +3,40 @@ import BellStrokeIcon from "@artsy/icons/BellStrokeIcon"
 }
 
-export const CreateAlertButton: FC<Props> = ({ onClick, ...buttonProps }) => {
+export interface Props extends ButtonProps {
```
`CreateAlertButtonProps`

damassi commented on src/Components/Alert/Hooks/__tests__/AlertProvider.jest.tsx:
```diff
@@ -0,0 +1,122 @@
+  let mockUseTracking: jest.Mock = useTracking as jest.Mock
+  let mockUseSystemContext: jest.Mock = useSystemContext as jest.Mock
+
+  const { getWrapper } = setupTestWrapper({
```
We should be using RTL, not enzyme (`setupTestWrapperTL`)

dblandin replied:
🤦🏼‍♂️ Thanks! Will update.

damassi commented on src/Components/Alert/Hooks/__tests__/AlertProvider.jest.tsx:
```diff
@@ -0,0 +1,122 @@
+      {},
+      relayEnv
+    )
+    let page = new AlertProviderTestPage(wrapper)
```
The `Page` pattern is something we've been refactoring out of the app, so we shouldn't use that here in favor of vanilla testing-library/react assertions

damassi commented on src/Components/Alert/Hooks/__tests__/Utils/AlertProviderTestPage.tsx:
```diff
@@ -0,0 +1,39 @@
+import { RootTestPage } from "DevTools/RootTestPage"
+
+export class AlertProviderTestPage extends RootTestPage {
```
I feel bad flagging this, because it seems fine, but yeah, this is a deprecated pattern

dblandin commented:
Addressed most of the feedback here so I'm going to go ahead and merge this in. Will continue with the migration to React Test Library over Enzyme in a follow-up PR.
